### PR TITLE
Fix stale closures in ref callbacks

### PR DESCRIPTION
**Pull request by HermanNygaard**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -98,6 +98,10 @@ export function useClickAway(cb) {
   const ref = React.useRef(null);
   const refCb = React.useRef(cb);
 
+  React.useLayoutEffect(() => {
```
Comment (Jeffzholy) [inline index.js:101]: Is it possible to assign `cb` directly to `refCb.current` in the top level of the hook instead of using `React.useLayoutEffect`?

Comment (HermanNygaard) [reply index.js:101]: No, as explained in the [docs](https://react.dev/reference/react/useRef#caveats): 
"Do not write or read ref.current during rendering, except for [initialization.](https://react.dev/reference/react/useRef#avoiding-recreating-the-ref-contents) This makes your component’s behavior unpredictable."

**Comment by tylermcginnis** [comment]
Thank you!

**Comment by AmitJoki** [comment]
Hi @HermanNygaard I was going through the source code of `useClickAway` and wasn't sure why `useLayoutEffect` was used when there's nothing "layout" related being done there.

I posted a question on StackOverflow https://stackoverflow.com/questions/79041810/why-is-uselayouteffect-used-in-useclickaway-react-hook

I have a couple of questions:

1. Why does the below not suffice?

```
React.useEffect(() => {
  refCb.current = cb;
}, [cb]);
```

2. Why is `cb` not in the `deps` of `useLayoutEffect`? Is it because it's very cheap to just assign it on every render without caring if `cb` has changed?

After going through the [link](https://github.com/facebook/react/issues/16956#issuecomment-536663843) under the question, I figured solving the stale value can be done earlier in `useLayoutEffect` than in `useEffect`. Is that rationale behind using `useLayoutEffect` here?

Thanks.

**Comment by jeppester** [comment]
I'm getting errors from inertia's SSR because it does not support `useLayoutEffect`.
While that is an issue that could be solved on inertia's side, my debugging effort lead me here.

I'm not sure I understand why there is a need to use a hook for assigning the callback.

If the goal is to have refCb.current always assigned to the latest callback, why not just assign `refCb.current` to `cb` in the root of the hook function?

**Comment by sagarpanchal** [comment]
> I'm getting errors from inertia's SSR because it does not support `useLayoutEffect`. While that is an issue that could be solved on inertia's side, my debugging effort lead me here.
> 
> I'm not sure I understand why there is a need to use a hook for assigning the callback.
> 
> If the goal is to have refCb.current always assigned to the latest callback, why not just assign `refCb.current` to `cb` in the root of the hook function?

@jeppester [useRef caveats](https://react.dev/reference/react/useRef#caveats:~:text=Do%20not%20write%20or%20read%20ref.current%20during%20rendering%2C%20except%20for%20initialization.%20This%20makes%20your%20component%E2%80%99s%20behavior%20unpredictable.)

**Pull request merged**